sympy/physics/continuum_mechanics/beam.py

```diff
@@ -0,0 +1,501 @@
+"""
+This module can be used to solve 2D beam bending problems with
+singularity functions in mechanics.
+
+"""
+
+from __future__ import print_function, division
+
+from sympy.core import S, Symbol, Eq, diff
+from sympy.solvers import linsolve
+from sympy.printing import sstr
+from sympy.functions import SingularityFunction
+from sympy.core import sympify
+from sympy.integrals import integrate
+
+
+class Beam(object):
+    """
+    A Beam is a structural element that is capable of withstanding load
+    primarily by resisting against bending.
+    Beams are characterized by their profile (Second moment of area),
+    their length, and their material.
+
+    Parameters
+    ==========
+    length : Sympifyable
+        A SymPy expression representing the Beam's length.
+    elastic_modulus : Sympifyable
+        A SymPy expression representing the Beam's Modulus of Elasticity.
+        It is a measure of the stiffness of the Beam material.
+    second_moment : Sympifyable
+        A SymPy expression representing the Beam's Second moment of area.
+        It is a geometrical property of an area which reflects how its
+        points are distributed with respect to its neutral axis.
+
+    Examples
+    ========
+    >>> from sympy.physics.continuum_mechanics.beam import Beam
+    >>> from sympy import Symbol
+    >>> E = Symbol('E')
+    >>> I = Symbol('I')
+    >>> Beam(1, E, I)
+    Beam(1, E, I)
+
+    """
+
+    def __init__(self, length, elastic_modulus, second_moment):
+        self._length = length
+        self._elastic_modulus = elastic_modulus
+        self._second_moment = second_moment
+        self._boundary_conditions = {'deflection': [], 'moment': [], 'slope': []}
+        self._load = 0
+
+    def __str__(self):
+        str_sol = 'Beam(%s, %s, %s)' % (sstr(self._length), sstr(self._elastic_modulus), sstr(self._second_moment))
+        return str_sol
+
+    __repr__ = __str__
+
+    @property
+    def length(self):
+        """Length of the Beam."""
+        return self._length
+
+    @length.setter
+    def length(self, l):
+        self._length = sympify(l)
+
+    @property
+    def elastic_modulus(self):
+        """Young's Modulus of the Beam. """
+        return self._elastic_modulus
+
+    @elastic_modulus.setter
+    def elastic_modulus(self, e):
+        self._elastic_modulus = sympify(e)
+
+    @property
+    def second_moment(self):
+        """Second moment of area of the Beam. """
+        return self._second_moment
+
+    @second_moment.setter
+    def second_moment(self, i):
+        self._second_moment = sympify(i)
+
+    @property
+    def boundary_conditions(self):
+        """
+        Returns a dictionary of boundary conditions applied on the beam.
+        """
+        return self._boundary_conditions
+
+    def apply_boundary_conditions(self, **bcs):
```

Comment:
I feel that only this method should be allowed to change the `_boundary_conditions` dictionary. I mean:- 
```py
>>> b.apply_boundary_conditions(moment = [(0, 4), (4, 0)], deflection = [(0, 2)], slope = [(0, 1)])
>>> b.boundary_conditions
{'deflection': [(0, 2)], 'moment': [(0, 4), (4, 0)], 'slope': [(0, 1)]}
```
Now if user passes different values then this method shouldn't append them. It should change the whole list into a new list having only the passed values.
```py
>>> b.apply_boundary_conditions(moment = [(0, 4)], deflection = [(0, 4)], slope = [(0, 1), (4, 0)])
>>> b.boundary_conditions
{'deflection': [(0, 4)], 'moment': [(0, 4)], 'slope': [(0, 1), (4, 0)]}
 
```
Just like this.

But if the user wants to append some new boundary conditions with the existing ones, then they are free to use `apply_moment_boundary_conditions`, `apply_slope_boundary_conditions`, `apply_deflection_boundary_conditions`.

e.g.,
```py
>>> b.apply_boundary_conditions(moment = [(0, 4), (4, 0)], deflection = [(0, 2)], slope = [(0, 1)])

>>> b.boundary_conditions
{'deflection': [(0, 2)], 'moment': [(0, 4), (4, 0)], 'slope': [(0, 1)]}

>>> b.apply_moment_boundary_conditions((4, 3), (5, 0))
[(0, 4), (4, 0), (4, 3), (5, 0)]

>>> b.boundary_conditions
{'deflection': [(0, 2)], 'moment': [(0, 4), (4, 0), (4, 3), (5, 0)], 'slope': [(0, 1)]}

>>> b.apply_boundary_conditions(moment = [(0, 4)], deflection = [(0, 4)], slope = [(0, 1), (4, 0)])

>>> b.boundary_conditions
{'deflection': [(0, 4)], 'moment': [(0, 4)], 'slope': [(0, 1), (4, 0)]}
```
Any suggestions are welcome.